### PR TITLE
Parallelization for ROIpooling OP

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##

**What's the problem?**

ROIpooling is used in the Faster-RCNN. It will consume lots of time in the inference path because of the current implementation [(here)](https://github.com/apache/incubator-mxnet/blob/master/src/operator/roi_pooling.cc
) is not parallelized. 

One profiling results as below:

Time   of each OP: | ms | ms/call | calls
-- | -- | -- | --
ROIPooling | **38718.288** | 1548.73152 | 25
Convolution | 9963.628 | 3.724720748 | 2675
Reshape | 0.677 | 0.00677 | 100
Activation | 1089.6 | 0.427294118 | 2550
SoftmaxActivation | 43.279 | 1.73116 | 25
add_n | 1664.403 | 2.017458182 | 825
Pooling | 68.531 | 1.37062 | 50
_contrib_Proposal | 351.051 | 14.04204 | 25
softmax | 0.658 | 0.02632 | 25
FullyConnected | 29.328 | 0.58656 | 50
BatchNorm | 2865.261 | 1.123631765 | 2550

**What we have tried**

We @pengzhao-intel @TaoLv have parallelized this pooling algorithm and got the 20+X performance improvement by OpenMP directives.

Faster-RCNN

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Parallelization for ROIpooling

### CODE REVIEW

File: src/operator/roi_pooling.cc [hunk 1]
```diff
@@ -74,7 +76,13 @@ inline void ROIPoolForward(const Tensor<cpu, 4, Dtype> &out,
 
     const Dtype* batch_data = bottom_data + data_size * roi_batch_ind;
 
+    #pragma omp parallel for firstprivate(batch_data, top_data, argmax_data)
     for (int c = 0; c < channels_; ++c) {
+      // Increment all data pointers
+      batch_data += c * data.size(2) * data.size(3);
```
Comment (cjolivier01) [inline src/operator/roi_pooling.cc]: please cache the xx.size(2) * you.size(3) result outside of the loops since they are invariant, so as to perform fewer multiplies.

Comment (cjolivier01) [reply src/operator/roi_pooling.cc]: if you do that, you can just keep adding that number each pass of c and then you don’t need to do a multiple for these at all

File: src/operator/roi_pooling.cc [hunk 2]
```diff
@@ -113,15 +121,19 @@ inline void ROIPoolForward(const Tensor<cpu, 4, Dtype> &out,
           }
         }
       }
-      // Increment all data pointers by one channel
-      batch_data += data.size(2) * data.size(3);
-      top_data += out.size(2) * out.size(3);
-      argmax_data += max_idx.size(2) * max_idx.size(3);
+      // Decrement all data pointers
+      batch_data -= c * data.size(2) * data.size(3);
```
Comment (cjolivier01) [inline src/operator/roi_pooling.cc]: same here but subtract

File: src/operator/roi_pooling.cc (hunk 1 above)
Comment (cjolivier01) [reply src/operator/roi_pooling.cc]: multiply*

File: src/operator/roi_pooling.cc [hunk 3]
```diff
@@ -74,7 +79,13 @@ inline void ROIPoolForward(const Tensor<cpu, 4, Dtype> &out,
 
     const Dtype* batch_data = bottom_data + data_size * roi_batch_ind;
 
+    #pragma omp parallel for firstprivate(batch_data, top_data, argmax_data)
```
Comment (cjolivier01) [inline src/operator/roi_pooling.cc]: curious why it’s not “parallel for” pragma?

Comment (TaoLv) [reply src/operator/roi_pooling.cc]: Do you mean why there is a `firstprivate` clause here?

Comment (cjolivier01) [reply src/operator/roi_pooling.cc]: Oh, I totally missed the 'for' in the pragma. Sorry.

File: src/operator/roi_pooling.cc [hunk 4]
```diff
@@ -98,30 +109,26 @@ inline void ROIPoolForward(const Tensor<cpu, 4, Dtype> &out,
...
       }
-      // Increment all data pointers by one channel
-      batch_data += data.size(2) * data.size(3);
-      top_data += out.size(2) * out.size(3);
-      argmax_data += max_idx.size(2) * max_idx.size(3);
     }
-    // Increment ROI data pointer
-    bottom_rois += bbox.size(1);
+    // Increase data pointers by one outsize
+    top_data += channels_ * out_size_c;
```
Comment (cjolivier01) [inline src/operator/roi_pooling.cc]: Can these muls be taken out of the loop?

File: src/operator/roi_pooling.cc [hunk 5]
```diff
@@ -74,7 +79,13 @@ inline void ROIPoolForward(const Tensor<cpu, 4, Dtype> &out,
 
     const Dtype* batch_data = bottom_data + data_size * roi_batch_ind;
 
+    #pragma omp parallel for firstprivate(batch_data, top_data, argmax_data)
     for (int c = 0; c < channels_; ++c) {
+      // Increment all data pointers
+      const Dtype* batch_data_c = batch_data + c * data_size_c;
+      Dtype* top_data_c = top_data + c * out_size_c;
+      Dtype* argmax_data_c = argmax_data + c * max_idx_size_c;
+
```
Comment (cjolivier01) [inline src/operator/roi_pooling.cc]: Nice work!
By the way, Multiplies are a lot slower than adds, so you can get even faster by reducing the muls in the loop ( + and * replace by single +):
```cpp
const Dtype *batch_data_c = batch_data;
Dtype *top_data_c = top_data, *argmax_data_c = argmax_data; 
for (int c = 0; c < channels_; ++c) {
  // Increment all data pointers
  batch_data_c  += data_size_c;
  top_data_c    += out_size_c;
  argmax_data_c += max_idx_size_c;
  .
  .
  .
}
```

Comment (cjolivier01) [reply src/operator/roi_pooling.cc]: In other places too...

Comment (cjolivier01) [reply src/operator/roi_pooling.cc]: Sorry, I mean like this:
```cpp
const Dtype *batch_data_c = batch_data;
Dtype *top_data_c = top_data, *argmax_data_c = argmax_data;
for (int c = 0; c < channels_; ++c, 
                               batch_data_c  += data_size_c,
                               top_data_c    += out_size_c,
                               argmax_data_c += max_idx_size_c 
) {

}
```

Comment (pengzhao-intel) [reply src/operator/roi_pooling.cc]: @cjolivier01 Thanks for the great comments to Xinyu.
Regarding replacing multiplication (FMA) to incremental addition (+=), two points of my view:

1)  The incremental addition is more concise and logical rather than computing the index from start point each time by multiplication. But there're strong compute dependency and we **CANT** start the loop N+1 before the loop N. 
Thus, we have to change to multiple styles for the parallelization.

2) The efficiency of Multiplication (FMA) and addition (ADD) is same in the latest HW (Intel skylake)
Take SSE instruct as an example:
ADD, latency 4, CPI 0.5;
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=add&techs=SSE&expand=127,3673,127
FMA (MUL+ADD), latency 4, CPI 0.5
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=127,3673,2395,3673,2395,2407&text=mul&techs=FMA,Other

File: src/operator/roi_pooling.cc (hunk 4 above)
Comment (xinyu-intel) [reply src/operator/roi_pooling.cc]: Thanks for great comment, I've redesigned this part:)

**Comment by piiswrong** [comment]
@cjolivier01

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!

**Comment by pengzhao-intel** [comment]
@cjolivier01 Could you help review again? 
Our customer requested this features and want to get the usable code from official repo ASAP.
Thanks in advance.

File: src/operator/roi_pooling.cc (hunk 5 above)
Comment (cjolivier01) [reply src/operator/roi_pooling.cc]: ok

**Comment by cjolivier01** [comment]
LGTM

File: src/operator/roi_pooling.cc (hunk 5 above)
Comment (cjolivier01) [reply src/operator/roi_pooling.cc]: try to keep in mind in the future that we support less-awesome CPUs than Intel Skylake :)

**Comment by piiswrong** [comment]
@cjolivier01 
Shouldn't the omp pragma get number of threads from OpenMP::Get()?

**Comment by cjolivier01** [comment]
it probably wouldn’t hurt. I don’t think it’s critical, though since channels tends to be a small number and it doesn’t look like the intent is to not use OMP if a GPU was used. it’s sort of like batch norm in this way.
but i suppose
for the sake of consistency it maybe should be fixed
